File: ListSorter.java
// ListSorter.java - Сортировщик списков на Java (CLI)
import java.io.*;
import java.nio.file.*;
import java.util.*;
import java.time.*;

public class ListSorter {
    private static final Scanner scanner = new Scanner(System.in);
    private static final Random random = new Random();

    interface SortFunc {
        List<Integer> sort(List<Integer> arr);
    }

    // ========== АЛГОРИТМЫ СОРТИРОВКИ ==========
    static class Algorithms {
        static List<Integer> bubbleSort(List<Integer> arr) {
            List<Integer> result = new ArrayList<>(arr);
            int n = result.size();
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < n - i - 1; j++) {
                    if (result.get(j) > result.get(j + 1)) {
                        Collections.swap(result, j, j + 1);
                    }
                }
            }
            return result;
        }

        static List<Integer> quickSort(List<Integer> arr) {
            if (arr.size() <= 1) return new ArrayList<>(arr);
            int pivot = arr.get(arr.size() / 2);
            List<Integer> left = new ArrayList<>();
            List<Integer> middle = new ArrayList<>();
            List<Integer> right = new ArrayList<>();
            for (int x : arr) {
                if (x < pivot) left.add(x);
                else if (x == pivot) middle.add(x);
                else right.add(x);
            }
            List<Integer> result = quickSort(left);
            result.addAll(middle);
            result.addAll(quickSort(right));
            return result;
        }

        static List<Integer> mergeSort(List<Integer> arr) {
            if (arr.size() <= 1) return new ArrayList<>(arr);
            int mid = arr.size() / 2;
            List<Integer> left = mergeSort(arr.subList(0, mid));
            List<Integer> right = mergeSort(arr.subList(mid, arr.size()));
            return merge(left, right);
        }

        static List<Integer> merge(List<Integer> left, List<Integer> right) {
            List<Integer> result = new ArrayList<>();
            int i = 0, j = 0;
            while (i < left.size() && j < right.size()) {
                if (left.get(i) <= right.get(j)) {
                    result.add(left.get(i++));
                } else {
                    result.add(right.get(j++));
                }
            }
            result.addAll(left.subList(i, left.size()));
            result.addAll(right.subList(j, right.size()));
            return result;
        }

        static List<Integer> selectionSort(List<Integer> arr) {
            List<Integer> result = new ArrayList<>(arr);
            int n = result.size();
            for (int i = 0; i < n; i++) {
                int minIdx = i;
                for (int j = i + 1; j < n; j++) {
                    if (result.get(j) < result.get(minIdx)) minIdx = j;
                }
                Collections.swap(result, i, minIdx);
            }
            return result;
        }

        static List<Integer> insertionSort(List<Integer> arr) {
            List<Integer> result = new ArrayList<>(arr);
            for (int i = 1; i < result.size(); i++) {
                int key = result.get(i);
                int j = i - 1;
                while (j >= 0 && result.get(j) > key) {
                    result.set(j + 1, result.get(j));
                    j--;
                }
                result.set(j + 1, key);
            }
            return result;
        }

        static List<Integer> builtinSort(List<Integer> arr) {
            List<Integer> result = new ArrayList<>(arr);
            Collections.sort(result);
            return result;
        }
    }

    static class Algorithm {
        String name;
        SortFunc func;
        Algorithm(String name, SortFunc func) {
            this.name = name;
            this.func = func;
        }
    }

    private static final Map<String, Algorithm> ALGORITHMS = new LinkedHashMap<>();
    static {
        ALGORITHMS.put("1", new Algorithm("Пузырьковая", Algorithms::bubbleSort));
        ALGORITHMS.put("2", new Algorithm("Быстрая", Algorithms::quickSort));
        ALGORITHMS.put("3", new Algorithm("Слиянием", Algorithms::mergeSort));
        ALGORITHMS.put("4", new Algorithm("Стандартная", Algorithms::builtinSort));
        ALGORITHMS.put("5", new Algorithm("Выбором", Algorithms::selectionSort));
        ALGORITHMS.put("6", new Algorithm("Вставками", Algorithms::insertionSort));
    }

    // ========== ВСПОМОГАТЕЛЬНЫЕ ФУНКЦИИ ==========
    static List<Integer> generateRandomList(int size, int minVal, int maxVal) {
        List<Integer> list = new ArrayList<>();
        for (int i = 0; i < size; i++) {
            list.add(random.nextInt(maxVal - minVal + 1) + minVal);
        }
        return list;
    }

    static Map.Entry<List<Integer>, Double> measureTime(SortFunc func, List<Integer> data) {
        List<Integer> arr = new ArrayList<>(data);
        Instant start = Instant.now();
        List<Integer> result = func.sort(arr);
        double elapsed = Duration.between(start, Instant.now()).toNanos() / 1_000_000_000.0;
        return Map.entry(result, elapsed);
    }

    static void printTable(Map<String, Map.Entry<List<Integer>, Double>> results) {
        if (results.isEmpty()) {
            System.out.println("Нет данных.");
            return;
        }
        double maxTime = results.values().stream().mapToDouble(Map.Entry::getValue).max().orElse(0);
        System.out.println("\n" + "=".repeat(70));
        System.out.printf("%70s\n", "РЕЗУЛЬТАТЫ СОРТИРОВКИ");
        System.out.println("=".repeat(70));
        for (Map.Entry<String, Map.Entry<List<Integer>, Double>> entry : results.entrySet()) {
            String name = entry.getKey();
            double elapsed = entry.getValue().getValue();
            int barLen = maxTime > 0 ? (int)((elapsed / maxTime) * 40) : 0;
            String bar = "█".repeat(barLen) + "░".repeat(40 - barLen);
            System.out.printf("%-15s %.6f сек.  %s\n", name, elapsed, bar);
        }
        System.out.println("=".repeat(70));
        if (!results.isEmpty()) {
            List<Integer> firstResult = results.values().iterator().next().getKey();
            if (firstResult.size() <= 20) {
                System.out.println("Отсортированный список: " + firstResult);
            } else {
                System.out.println("Отсортированный список (первые 20): " + firstResult.subList(0, 20));
            }
        }
    }

    static void exportCSV(Map<String, Map.Entry<List<Integer>, Double>> results, String filename) throws IOException {
        try (PrintWriter pw = new PrintWriter(filename)) {
            pw.println("Алгоритм,Время (сек),Размер списка");
            int size = results.values().iterator().next().getKey().size();
            for (Map.Entry<String, Map.Entry<List<Integer>, Double>> entry : results.entrySet()) {
                pw.printf("%s,%f,%d\n", entry.getKey(), entry.getValue().getValue(), size);
            }
        }
    }

    static void exportJSON(Map<String, Map.Entry<List<Integer>, Double>> results, String filename) throws IOException {
        StringBuilder sb = new StringBuilder();
        sb.append("{\n");
        boolean first = true;
        int size = results.values().iterator().next().getKey().size();
        for (Map.Entry<String, Map.Entry<List<Integer>, Double>> entry : results.entrySet()) {
            if (!first) sb.append(",\n");
            first = false;
            sb.append("  \"").append(entry.getKey()).append("\": {\n");
            sb.append("    \"sorted\": ").append(entry.getValue().getKey()).append(",\n");
            sb.append("    \"time\": ").append(entry.getValue().getValue()).append(",\n");
            sb.append("    \"size\": ").append(size).append("\n");
            sb.append("  }");
        }
        sb.append("\n}");
        Files.write(Paths.get(filename), sb.toString().getBytes());
    }

    static void interactive() throws IOException {
        System.out.println("📊 СОРТИРОВЩИК СПИСКОВ");
        while (true) {
            System.out.println("\nВыберите действие:");
            System.out.println("1. Сортировать введённый список");
            System.out.println("2. Сгенерировать случайный список");
            System.out.println("3. Сравнить все алгоритмы");
            System.out.println("0. Выход");
            System.out.print("Ваш выбор: ");
            String choice = scanner.nextLine().trim();
            if (choice.equals("0")) break;
            else if (choice.equals("1")) {
                System.out.print("Введите числа через пробел: ");
                String[] parts = scanner.nextLine().trim().split("\\s+");
                List<Integer> data = new ArrayList<>();
                for (String p : parts) {
                    try { data.add(Integer.parseInt(p)); } catch (NumberFormatException e) {}
                }
                if (data.isEmpty()) { System.out.println("Список пуст."); continue; }
                Map<String, Map.Entry<List<Integer>, Double>> results = new LinkedHashMap<>();
                System.out.println("\nВыберите алгоритм (или all для всех):");
                for (Map.Entry<String, Algorithm> entry : ALGORITHMS.entrySet()) {
                    System.out.println(entry.getKey() + ". " + entry.getValue().name);
                }
                System.out.print("Ваш выбор: ");
                String algoChoice = scanner.nextLine().trim();
                if (algoChoice.equals("all")) {
                    for (Algorithm alg : ALGORITHMS.values()) {
                        Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                        results.put(alg.name, res);
                    }
                } else if (ALGORITHMS.containsKey(algoChoice)) {
                    Algorithm alg = ALGORITHMS.get(algoChoice);
                    Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                    results.put(alg.name, res);
                } else {
                    System.out.println("Неверный выбор.");
                    continue;
                }
                printTable(results);
                System.out.print("Экспортировать результаты? (y/n): ");
                String export = scanner.nextLine().trim().toLowerCase();
                if (export.equals("y")) {
                    System.out.print("Формат (csv/json): ");
                    String fmt = scanner.nextLine().trim();
                    System.out.print("Имя файла: ");
                    String filename = scanner.nextLine().trim();
                    if (filename.isEmpty()) filename = "results." + fmt;
                    if (fmt.equals("csv")) exportCSV(results, filename);
                    else exportJSON(results, filename);
                    System.out.println("Экспортировано в " + filename);
                }
            } else if (choice.equals("2")) {
                System.out.print("Размер списка: ");
                int size = Integer.parseInt(scanner.nextLine().trim());
                System.out.print("Минимальное значение: ");
                int minVal = Integer.parseInt(scanner.nextLine().trim());
                System.out.print("Максимальное значение: ");
                int maxVal = Integer.parseInt(scanner.nextLine().trim());
                List<Integer> data = generateRandomList(size, minVal, maxVal);
                System.out.println("Сгенерированный список: " + data);
                Map<String, Map.Entry<List<Integer>, Double>> results = new LinkedHashMap<>();
                System.out.println("\nВыберите алгоритм (или all для всех):");
                for (Map.Entry<String, Algorithm> entry : ALGORITHMS.entrySet()) {
                    System.out.println(entry.getKey() + ". " + entry.getValue().name);
                }
                System.out.print("Ваш выбор: ");
                String algoChoice = scanner.nextLine().trim();
                if (algoChoice.equals("all")) {
                    for (Algorithm alg : ALGORITHMS.values()) {
                        Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                        results.put(alg.name, res);
                    }
                } else if (ALGORITHMS.containsKey(algoChoice)) {
                    Algorithm alg = ALGORITHMS.get(algoChoice);
                    Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                    results.put(alg.name, res);
                } else {
                    System.out.println("Неверный выбор.");
                    continue;
                }
                printTable(results);
            } else if (choice.equals("3")) {
                System.out.print("Введите числа через пробел (или оставьте пустым для случайных): ");
                String input = scanner.nextLine().trim();
                List<Integer> data;
                if (input.isEmpty()) {
                    System.out.print("Размер случайного списка: ");
                    int size = Integer.parseInt(scanner.nextLine().trim());
                    data = generateRandomList(size, 1, 100);
                    System.out.println("Сгенерированный список: " + data);
                } else {
                    data = new ArrayList<>();
                    for (String p : input.split("\\s+")) {
                        try { data.add(Integer.parseInt(p)); } catch (NumberFormatException e) {}
                    }
                }
                if (data.isEmpty()) { System.out.println("Список пуст."); continue; }
                Map<String, Map.Entry<List<Integer>, Double>> results = new LinkedHashMap<>();
                for (Algorithm alg : ALGORITHMS.values()) {
                    Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                    results.put(alg.name, res);
                }
                printTable(results);
                System.out.print("Экспортировать результаты? (y/n): ");
                String export = scanner.nextLine().trim().toLowerCase();
                if (export.equals("y")) {
                    System.out.print("Формат (csv/json): ");
                    String fmt = scanner.nextLine().trim();
                    System.out.print("Имя файла: ");
                    String filename = scanner.nextLine().trim();
                    if (filename.isEmpty()) filename = "results." + fmt;
                    if (fmt.equals("csv")) exportCSV(results, filename);
                    else exportJSON(results, filename);
                    System.out.println("Экспортировано в " + filename);
                }
            } else {
                System.out.println("Неверный выбор.");
            }
        }
    }

    public static void main(String[] args) throws IOException {
        if (args.length > 0) {
            String listFlag = null, algorithm = null, exportCsv = null, exportJson = null;
            boolean compare = false;
            int randomSize = 0;
            for (int i = 0; i < args.length; i++) {
                if (args[i].equals("--list")) {
                    listFlag = args[++i];
                } else if (args[i].equals("--random")) {
                    randomSize = Integer.parseInt(args[++i]);
                } else if (args[i].equals("--algorithm")) {
                    algorithm = args[++i];
                } else if (args[i].equals("--export-csv")) {
                    exportCsv = args[++i];
                } else if (args[i].equals("--export-json")) {
                    exportJson = args[++i];
                } else if (args[i].equals("--compare")) {
                    compare = true;
                }
            }
            List<Integer> data = new ArrayList<>();
            if (listFlag != null) {
                for (String p : listFlag.split(",")) {
                    try { data.add(Integer.parseInt(p.trim())); } catch (NumberFormatException e) {}
                }
            } else if (randomSize > 0) {
                data = generateRandomList(randomSize, 1, 100);
            } else {
                System.out.println("Укажите --list или --random");
                return;
            }
            if (data.isEmpty()) { System.out.println("Список пуст."); return; }
            Map<String, Map.Entry<List<Integer>, Double>> results = new LinkedHashMap<>();
            if (compare || "all".equals(algorithm)) {
                for (Algorithm alg : ALGORITHMS.values()) {
                    Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                    results.put(alg.name, res);
                }
            } else if (ALGORITHMS.containsKey(algorithm)) {
                Algorithm alg = ALGORITHMS.get(algorithm);
                Map.Entry<List<Integer>, Double> res = measureTime(alg.func, data);
                results.put(alg.name, res);
            } else {
                System.out.println("Укажите --algorithm или --compare");
                return;
            }
            printTable(results);
            if (exportCsv != null) { exportCSV(results, exportCsv); System.out.println("Экспортировано в " + exportCsv); }
            if (exportJson != null) { exportJSON(results, exportJson); System.out.println("Экспортировано в " + exportJson); }
        } else {
            interactive();
        }
    }
}

File: list_sorter.rs
// list_sorter.rs - Сортировщик списков на Rust (CLI)
use serde::{Serialize, Deserialize};
use std::collections::HashMap;
use std::fs;
use std::io::{self, Write, BufRead};
use std::time::Instant;
use rand::Rng;

// ========== АЛГОРИТМЫ СОРТИРОВКИ ==========
fn bubble_sort(arr: &[i32]) -> Vec<i32> {
    let mut result = arr.to_vec();
    let n = result.len();
    for i in 0..n {
        for j in 0..n - i - 1 {
            if result[j] > result[j + 1] {
                result.swap(j, j + 1);
            }
        }
    }
    result
}

fn quick_sort(arr: &[i32]) -> Vec<i32> {
    if arr.len() <= 1 {
        return arr.to_vec();
    }
    let pivot = arr[arr.len() / 2];
    let left: Vec<i32> = arr.iter().filter(|&&x| x < pivot).cloned().collect();
    let middle: Vec<i32> = arr.iter().filter(|&&x| x == pivot).cloned().collect();
    let right: Vec<i32> = arr.iter().filter(|&&x| x > pivot).cloned().collect();
    let mut result = quick_sort(&left);
    result.extend(middle);
    result.extend(quick_sort(&right));
    result
}

fn merge_sort(arr: &[i32]) -> Vec<i32> {
    if arr.len() <= 1 {
        return arr.to_vec();
    }
    let mid = arr.len() / 2;
    let left = merge_sort(&arr[..mid]);
    let right = merge_sort(&arr[mid..]);
    merge(&left, &right)
}

fn merge(left: &[i32], right: &[i32]) -> Vec<i32> {
    let mut result = Vec::with_capacity(left.len() + right.len());
    let mut i = 0;
    let mut j = 0;
    while i < left.len() && j < right.len() {
        if left[i] <= right[j] {
            result.push(left[i]);
            i += 1;
        } else {
            result.push(right[j]);
            j += 1;
        }
    }
    result.extend_from_slice(&left[i..]);
    result.extend_from_slice(&right[j..]);
    result
}

fn selection_sort(arr: &[i32]) -> Vec<i32> {
    let mut result = arr.to_vec();
    let n = result.len();
    for i in 0..n {
        let mut min_idx = i;
        for j in i + 1..n {
            if result[j] < result[min_idx] {
                min_idx = j;
            }
        }
        result.swap(i, min_idx);
    }
    result
}

fn insertion_sort(arr: &[i32]) -> Vec<i32> {
    let mut result = arr.to_vec();
    for i in 1..result.len() {
        let key = result[i];
        let mut j = i as isize - 1;
        while j >= 0 && result[j as usize] > key {
            result[(j + 1) as usize] = result[j as usize];
            j -= 1;
        }
        result[(j + 1) as usize] = key;
    }
    result
}

fn builtin_sort(arr: &[i32]) -> Vec<i32> {
    let mut result = arr.to_vec();
    result.sort();
    result
}

type SortFunc = fn(&[i32]) -> Vec<i32>;

#[derive(Clone)]
struct Algorithm {
    name: String,
    func: SortFunc,
}

lazy_static::lazy_static! {
    static ref ALGORITHMS: HashMap<String, Algorithm> = {
        let mut m = HashMap::new();
        m.insert("1".to_string(), Algorithm { name: "Пузырьковая".to_string(), func: bubble_sort });
        m.insert("2".to_string(), Algorithm { name: "Быстрая".to_string(), func: quick_sort });
        m.insert("3".to_string(), Algorithm { name: "Слиянием".to_string(), func: merge_sort });
        m.insert("4".to_string(), Algorithm { name: "Стандартная".to_string(), func: builtin_sort });
        m.insert("5".to_string(), Algorithm { name: "Выбором".to_string(), func: selection_sort });
        m.insert("6".to_string(), Algorithm { name: "Вставками".to_string(), func: insertion_sort });
        m
    };
}

#[derive(Serialize, Deserialize)]
struct SortResult {
    sorted: Vec<i32>,
    time: f64,
    size: usize,
}

fn measure_time(func: SortFunc, data: &[i32]) -> (Vec<i32>, f64) {
    let start = Instant::now();
    let result = func(data);
    let elapsed = start.elapsed().as_secs_f64();
    (result, elapsed)
}

fn generate_random_list(size: usize, min_val: i32, max_val: i32) -> Vec<i32> {
    let mut rng = rand::thread_rng();
    (0..size).map(|_| rng.gen_range(min_val..=max_val)).collect()
}

fn print_table(results: &HashMap<String, (Vec<i32>, f64)>) {
    if results.is_empty() {
        println!("Нет данных.");
        return;
    }
    let max_time = results.values().map(|(_, t)| *t).fold(0.0, f64::max);
    println!("\n{}", "=".repeat(70));
    println!("{:^70}", "РЕЗУЛЬТАТЫ СОРТИРОВКИ");
    println!("{}", "=".repeat(70));
    for (name, (_, elapsed)) in results {
        let bar_len = if max_time > 0.0 { (elapsed / max_time * 40.0) as usize } else { 0 };
        let bar = "█".repeat(bar_len) + &"░".repeat(40 - bar_len);
        println!("{:<15} {:.6} сек.  {}", name, elapsed, bar);
    }
    println!("{}", "=".repeat(70));
    if let Some((_, (sorted, _))) = results.iter().next() {
        if sorted.len() <= 20 {
            println!("Отсортированный список: {:?}", sorted);
        } else {
            println!("Отсортированный список (первые 20): {:?}", &sorted[..20]);
        }
    }
}

fn export_csv(results: &HashMap<String, (Vec<i32>, f64)>, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut writer = csv::Writer::from_path(filename)?;
    writer.write_record(&["Алгоритм", "Время (сек)", "Размер списка"])?;
    let size = results.values().next().map(|(v, _)| v.len()).unwrap_or(0);
    for (name, (_, elapsed)) in results {
        writer.serialize((name, elapsed, size))?;
    }
    writer.flush()?;
    Ok(())
}

fn export_json(results: &HashMap<String, (Vec<i32>, f64)>, filename: &str) -> Result<(), Box<dyn std::error::Error>> {
    let mut data = HashMap::new();
    for (name, (sorted, time)) in results {
        data.insert(name.clone(), SortResult {
            sorted: sorted.clone(),
            time: *time,
            size: sorted.len(),
        });
    }
    let json = serde_json::to_string_pretty(&data)?;
    fs::write(filename, json)?;
    Ok(())
}

fn read_line(prompt: &str) -> String {
    print!("{}", prompt);
    io::stdout().flush().unwrap();
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    input.trim().to_string()
}

fn parse_numbers(s: &str) -> Vec<i32> {
    s.split_whitespace()
        .filter_map(|x| x.parse::<i32>().ok())
        .collect()
}

fn interactive() {
    println!("📊 СОРТИРОВЩИК СПИСКОВ");
    loop {
        println!("\nВыберите действие:");
        println!("1. Сортировать введённый список");
        println!("2. Сгенерировать случайный список");
        println!("3. Сравнить все алгоритмы");
        println!("0. Выход");
        let choice = read_line("Ваш выбор: ");
        match choice.as_str() {
            "0" => break,
            "1" => {
                let input = read_line("Введите числа через пробел: ");
                let data = parse_numbers(&input);
                if data.is_empty() {
                    println!("Список пуст.");
                    continue;
                }
                let mut results = HashMap::new();
                println!("\nВыберите алгоритм (или all для всех):");
                for (key, alg) in ALGORITHMS.iter() {
                    println!("{}. {}", key, alg.name);
                }
                let algo_choice = read_line("Ваш выбор: ");
                if algo_choice == "all" {
                    for (_, alg) in ALGORITHMS.iter() {
                        let (sorted, elapsed) = measure_time(alg.func, &data);
                        results.insert(alg.name.clone(), (sorted, elapsed));
                    }
                } else if let Some(alg) = ALGORITHMS.get(&algo_choice) {
                    let (sorted, elapsed) = measure_time(alg.func, &data);
                    results.insert(alg.name.clone(), (sorted, elapsed));
                } else {
                    println!("Неверный выбор.");
                    continue;
                }
                print_table(&results);
                let export = read_line("Экспортировать результаты? (y/n): ");
                if export.to_lowercase() == "y" {
                    let fmt = read_line("Формат (csv/json): ");
                    let filename = read_line("Имя файла: ");
                    let filename = if filename.is_empty() { format!("results.{}", fmt) } else { filename };
                    let res = if fmt == "csv" {
                        export_csv(&results, &filename)
                    } else {
                        export_json(&results, &filename)
                    };
                    if let Err(e) = res {
                        println!("Ошибка: {}", e);
                    } else {
                        println!("Экспортировано в {}", filename);
                    }
                }
            }
            "2" => {
                let size = read_line("Размер списка: ").parse::<usize>().unwrap_or(10);
                let min_val = read_line("Минимальное значение: ").parse::<i32>().unwrap_or(1);
                let max_val = read_line("Максимальное значение: ").parse::<i32>().unwrap_or(100);
                let data = generate_random_list(size, min_val, max_val);
                println!("Сгенерированный список: {:?}", data);
                let mut results = HashMap::new();
                println!("\nВыберите алгоритм (или all для всех):");
                for (key, alg) in ALGORITHMS.iter() {
                    println!("{}. {}", key, alg.name);
                }
                let algo_choice = read_line("Ваш выбор: ");
                if algo_choice == "all" {
                    for (_, alg) in ALGORITHMS.iter() {
                        let (sorted, elapsed) = measure_time(alg.func, &data);
                        results.insert(alg.name.clone(), (sorted, elapsed));
                    }
                } else if let Some(alg) = ALGORITHMS.get(&algo_choice) {
                    let (sorted, elapsed) = measure_time(alg.func, &data);
                    results.insert(alg.name.clone(), (sorted, elapsed));
                } else {
                    println!("Неверный выбор.");
                    continue;
                }
                print_table(&results);
            }
            "3" => {
                let input = read_line("Введите числа через пробел (или оставьте пустым для случайных): ");
                let data = if input.is_empty() {
                    let size = read_line("Размер случайного списка: ").parse::<usize>().unwrap_or(10);
                    let list = generate_random_list(size, 1, 100);
                    println!("Сгенерированный список: {:?}", list);
                    list
                } else {
                    parse_numbers(&input)
                };
                if data.is_empty() {
                    println!("Список пуст.");
                    continue;
                }
                let mut results = HashMap::new();
                for (_, alg) in ALGORITHMS.iter() {
                    let (sorted, elapsed) = measure_time(alg.func, &data);
                    results.insert(alg.name.clone(), (sorted, elapsed));
                }
                print_table(&results);
                let export = read_line("Экспортировать результаты? (y/n): ");
                if export.to_lowercase() == "y" {
                    let fmt = read_line("Формат (csv/json): ");
                    let filename = read_line("Имя файла: ");
                    let filename = if filename.is_empty() { format!("results.{}", fmt) } else { filename };
                    let res = if fmt == "csv" {
                        export_csv(&results, &filename)
                    } else {
                        export_json(&results, &filename)
                    };
                    if let Err(e) = res {
                        println!("Ошибка: {}", e);
                    } else {
                        println!("Экспортировано в {}", filename);
                    }
                }
            }
            _ => println!("Неверный выбор."),
        }
    }
}

fn main() {
    let args: Vec<String> = std::env::args().collect();
    if args.len() > 1 {
        let mut list = Vec::new();
        let mut random_size = 0;
        let mut algorithm = String::new();
        let mut export_csv = String::new();
        let mut export_json = String::new();
        let mut compare = false;
        let mut i = 1;
        while i < args.len() {
            match args[i].as_str() {
                "--list" => {
                    while i + 1 < args.len() && !args[i+1].starts_with("--") {
                        if let Ok(n) = args[i+1].parse::<i32>() {
                            list.push(n);
                        }
                        i += 1;
                    }
                }
                "--random" => {
                    random_size = args[i+1].parse().unwrap_or(0);
                    i += 1;
                }
                "--algorithm" => {
                    algorithm = args[i+1].clone();
                    i += 1;
                }
                "--export-csv" => {
                    export_csv = args[i+1].clone();
                    i += 1;
                }
                "--export-json" => {
                    export_json = args[i+1].clone();
                    i += 1;
                }
                "--compare" => {
                    compare = true;
                }
                _ => {}
            }
            i += 1;
        }
        let data = if !list.is_empty() {
            list
        } else if random_size > 0 {
            generate_random_list(random_size, 1, 100)
        } else {
            println!("Укажите --list или --random");
            return;
        };
        let mut results = HashMap::new();
        if compare || algorithm == "all" {
            for (_, alg) in ALGORITHMS.iter() {
                let (sorted, elapsed) = measure_time(alg.func, &data);
                results.insert(alg.name.clone(), (sorted, elapsed));
            }
        } else if let Some(alg) = ALGORITHMS.get(&algorithm) {
            let (sorted, elapsed) = measure_time(alg.func, &data);
            results.insert(alg.name.clone(), (sorted, elapsed));
        } else {
            println!("Укажите --algorithm или --compare");
            return;
        }
        print_table(&results);
        if !export_csv.is_empty() {
            if let Err(e) = export_csv(&results, &export_csv) {
                println!("Ошибка: {}", e);
            } else {
                println!("Экспортировано в {}", export_csv);
            }
        }
        if !export_json.is_empty() {
            if let Err(e) = export_json(&results, &export_json) {
                println!("Ошибка: {}", e);
            } else {
                println!("Экспортировано в {}", export_json);
            }
        }
    } else {
        interactive();
    }
}
